src/python/lib/ensembl/compara/citest/testdb.py

```diff
@@ -0,0 +1,194 @@
+
+        Note:
+            This method is called when ``self.runtest()`` raises an exception.
+
+        Args:
+            excinfo: Exception information with additional support for navigating and traceback.
+            style: Traceback print mode (``auto``/``long``/``short``/``line``/``native``/``no``).
+
+        """
+        if isinstance(excinfo.value, FailedDBTestException):
+            self.error_info['expected'] = excinfo.value.args[0]
+            self.error_info['found'] = excinfo.value.args[1]
+            self.error_info['query'] = str(excinfo.value.args[2]).replace('\n', '').strip()
+            return excinfo.value.args[3] + "\n"
+        if isinstance(excinfo.value, AssertionError):
+            return excinfo.value.args[0] + "\n"
+        return super().repr_failure(excinfo, style)
+
+    def get_report_header(self) -> str:
+        """Returns the header to display in the error report."""
+        return f"Database table: {self.table}, test: {self.name}"
+
+    def test_num_rows(self, variation: float = 0.0, group_by: Union[str, List] = None,
+                      filter_by: Union[str, List] = None) -> None:
+        """Compares the number of rows between reference and target tables.
+
+        If `group_by` is provided, the number of rows will be compared per group, applying the same variation
+        to all of them. If `filter_by` is provided, only the rows matching all the given conditions will be
+        compared.
+
+        Args:
+            variation: Allowed variation between reference and target tables.
+            group_by: Group rows by column(s), and count the number of rows per group.
+            filter_by: Filter rows by one or more conditions (joined by the AND operator).
+
+        Raise:
+            FailedDBTestException: If `group_by` is provided and the number of groups is different; or if the
+                number of rows differ for at least one group.
+
+        """
+        # Compose the sql query from the given parameters (both databases should have the same table schema)
+        table = self.ref_db.tables[self.table]
+        columns = [table.columns[col] for col in to_list(group_by)]
+        # Use primary key in count to improve the query performance
+        primary_key = self.ref_db.get_primary_key_columns(self.table)[0]
+        query = select(columns + [func.count(table.columns[primary_key]).label('nrows')])
+        if columns:
+            # ORDER BY to ensure that the results are always in the same order (for the same groups)
+            query = query.group_by(*columns).order_by(*columns)
+        for clause in to_list(filter_by):
+            query = query.where(text(clause))
+        # Get the number of rows for both databases
+        ref_data = pandas.read_sql(query, self.ref_db.connect())
+        target_data = pandas.read_sql(query, self.target_db.connect())
+        # Check if the size of the returned tables are the same
+        if ref_data.shape != target_data.shape:
```

Comment:
I thought about this, but this test only does count the number of rows (so it would pass as the number of rows are the same). For this case if you fear that may happen, you should use `test_content`. In my opinion, testing the number of rows works really well with eHive tables (grouping by `analysis_id`, for instance), but I would definitely use `test_content` for "compara" tables.

Reply:
That's fine if you don't have a GROUP BY. But if you have a GROUP BY, the test should be comparing the number rows between equivalent groupings. It doesn't make sense to compare the number of rows of genome_db_id 3 vs genome_db_id 4

Reply:
I see what you mean... And you are right. I will fix it now.